django/db/models/fields/related_descriptors.py

```diff
@@ -1107,14 +1107,26 @@ def get_prefetch_queryset(self, instances, queryset=None):
             join_table = fk.model._meta.db_table
             connection = connections[queryset.db]
             qn = connection.ops.quote_name
-            queryset = queryset.extra(
-                select={
-                    "_prefetch_related_val_%s"
-                    % f.attname: "%s.%s"
-                    % (qn(join_table), qn(f.column))
+            if (
+                connection.features.has_native_uuid_field
+                and "ForeignKey" == fk.__class__.__name__
+            ):
+                annotated_fields = {
+                    f"_prefetch_related_val_{f.attname}": F(
+                        f"{self.query_field_name}__{f.target_field.name}"
+                    )
                     for f in fk.local_related_fields
                 }
-            )
+                queryset = queryset.annotate(**annotated_fields)
+            else:
+                queryset = queryset.extra(
+                    select={
+                        "_prefetch_related_val_%s"
+                        % f.attname: "%s.%s"
+                        % (qn(join_table), qn(f.column))
+                        for f in fk.local_related_fields
+                    }
+                )
```

Comment:
@felixxm do you know what test(s) cover this functionality? If there is a test that cover this part of code and a test is correct then it is tested.

Reply:
this part of code (method `get_prefetch_queryset`) is covered by 74 existing tests so it means that `f.get_db_prep_value()` exists on fetched results.

Reply:
@raydeal It works on PostgreSQL because `psycopg` recognizes and returns `uuid.UUID`. On MariaDB strings are returned for `UUIDField` and `get_db_prep_value()` is necessary on related fields, as I [proposed](https://github.com/django/django/pull/16194/files#r1119736141). `UUIDField.get_db_prep_value()` is a no-op on PostgreSQL so adding `f.get_db_prep_value()` for local related fields shouldn't have any side-effects.

Reply:
@raydeal With this change we can revert others:
```diff
diff --git a/django/db/models/fields/related_descriptors.py b/django/db/models/fields/related_descriptors.py
index 7e9d581a0e..3f4e850b55 100644
--- a/django/db/models/fields/related_descriptors.py
+++ b/django/db/models/fields/related_descriptors.py
@@ -73,7 +73,7 @@ from django.db import (
     router,
     transaction,
 )
-from django.db.models import F, Q, Window, signals
+from django.db.models import Q, Window, signals
 from django.db.models.functions import RowNumber
 from django.db.models.lookups import GreaterThan, LessThanOrEqual
 from django.db.models.query import QuerySet
@@ -1107,26 +1107,14 @@ def create_forward_many_to_many_manager(superclass, rel, reverse):
             join_table = fk.model._meta.db_table
             connection = connections[queryset.db]
             qn = connection.ops.quote_name
-            if (
-                connection.features.has_native_uuid_field
-                and fk.get_internal_type() == "ForeignKey"
-            ):
-                annotated_fields = {
-                    f"_prefetch_related_val_{f.attname}": F(
-                        f"{self.query_field_name}__{f.target_field.name}"
-                    )
+            queryset = queryset.extra(
+                select={
+                    "_prefetch_related_val_%s"
+                    % f.attname: "%s.%s"
+                    % (qn(join_table), qn(f.column))
                     for f in fk.local_related_fields
                 }
-                queryset = queryset.annotate(**annotated_fields)
-            else:
-                queryset = queryset.extra(
-                    select={
-                        "_prefetch_related_val_%s"
-                        % f.attname: "%s.%s"
-                        % (qn(join_table), qn(f.column))
-                        for f in fk.local_related_fields
-                    }
-                )
+            )
             return (
                 queryset,
                 lambda result: tuple(

```

Reply:
you are right, thanks. Done.